File: app/src/main/java/br/com/alura/javapop/ui/activity/PullRequestActivity.java
package br.com.alura.javapop.ui.activity;

import android.app.PendingIntent;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import java.util.List;

import br.com.alura.javapop.R;
import br.com.alura.javapop.model.PullRequest;
import br.com.alura.javapop.model.Repositorio;
import br.com.alura.javapop.retrofit.RetrofitInicializador;
import br.com.alura.javapop.ui.adapter.ListaJavaPopAdapter;
import br.com.alura.javapop.ui.adapter.ListaPullRequestAdapter;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static br.com.alura.javapop.ui.constants.REPOSITORIO;
import static br.com.alura.javapop.ui.constants.REQUEST_CODE;

public class PullRequestActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ListaPullRequestAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pull_request);

        Repositorio repositorio = (Repositorio) getIntent().getSerializableExtra(REPOSITORIO);

        String nomeRepositorio = repositorio.getNome();
        String nomeUsuario = repositorio.getUsuario().getNome();

        setTitle(nomeRepositorio);

        Call<List<PullRequest>> call = new RetrofitInicializador().getAlunoService()
                .listaPullRequest(nomeUsuario,nomeRepositorio);

        call.enqueue(new Callback<List<PullRequest>>() {
            @Override
            public void onResponse(Call<List<PullRequest>>call, Response<List<PullRequest>> response) {
                List<PullRequest> pullRequests = response.body();
                configurarRecyclerView(pullRequests);
            }

            @Override
            public void onFailure(Call<List<PullRequest>> call, Throwable t) {
                Log.e("onFailure chamado", t.getMessage());

            }

        });


    }

    private void configurarRecyclerView(List<PullRequest> pullRequests) {
        recyclerView = findViewById(R.id.pull_request_recyclerview);
        LinearLayoutManager manager = configurarManager(recyclerView);
        configuraAdapter(pullRequests, recyclerView);
    }

    private LinearLayoutManager configurarManager(RecyclerView recyclerView) {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);
        return linearLayoutManager;
    }

    private void configuraAdapter(List<PullRequest> pullRequests, RecyclerView recyclerView) {
        adapter = new ListaPullRequestAdapter(this, pullRequests, selecionaPullRequest());
        recyclerView.setAdapter(adapter);
    }

    @NonNull
    private ListaPullRequestAdapter.OnItemClickListener selecionaPullRequest() {
        return new ListaPullRequestAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(PullRequest pullRequest) {
                Intent abrirNavegador = new Intent(Intent.ACTION_VIEW);
                abrirNavegador.setData(Uri.parse(pullRequest.getUrl()));
                startActivity(abrirNavegador);
            }
        };
    }
}

File: app/src/main/java/cl/carteaga/querygithub/PullRequestActivity.kt
package cl.carteaga.querygithub

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ProgressBar
import cl.carteaga.querygithub.classes.*
import cl.carteaga.querygithub.classes.responseJson.ResponsePullRequest
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PullRequestActivity : AppCompatActivity() {
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapterPullRequest: AdapterPullRequest
    private lateinit var viewManager: RecyclerView.LayoutManager
    private lateinit var progressBar: ProgressBar
    private var pullRequestService: GitHubEndPoints? = null
    private var nameRepository: String = ""
    private var nameAuthorRepository: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pull_request)

        val extras: Bundle = this.intent.extras
        nameRepository = extras["nameRepository"] as String
        nameAuthorRepository = extras["authorName"] as String

        this.title = nameRepository
        this.supportActionBar?.setDisplayHomeAsUpEnabled(true)

        progressBar = findViewById(R.id.progressBarPullRequest)
        recyclerView = findViewById(R.id.rcPullRequest)
        recyclerView.setHasFixedSize(true)
        viewManager = LinearLayoutManager(this)
        recyclerView.layoutManager = viewManager
        pullRequestService = GitHubApi.getClient()?.create(GitHubEndPoints::class.java)
        adapterPullRequest = AdapterPullRequest(mutableListOf())
        recyclerView.adapter = adapterPullRequest
        loadPage()
    }

    private fun loadPage() {
        progressBar.visibility = View.VISIBLE
        callPullRequestService()?.enqueue(object: Callback<List<ResponsePullRequest>> {
            override fun onFailure(call: Call<List<ResponsePullRequest>>,t: Throwable) {
            }

            override fun onResponse(call: Call<List<ResponsePullRequest>>,
                                    response: Response<List<ResponsePullRequest>>
            ) {
                if(response.code() == 200) {
                    var pullRequests = generatePullRequest(response.body())
                    adapterPullRequest = AdapterPullRequest(pullRequests)
                    recyclerView.adapter = adapterPullRequest
                    progressBar.visibility = View.GONE
                }
            }
        })
    }

    private fun callPullRequestService(): Call<List<ResponsePullRequest>>? {
        return pullRequestService?.getPullRequestRepository(nameAuthorRepository, nameRepository)
    }

    private fun generatePullRequest(items: List<ResponsePullRequest>?) : MutableList<PullRequest> {
        var pullrequests: MutableList<PullRequest> = mutableListOf()
        items?.map { item ->
            item.let {
                pullrequests.add(
                    PullRequest(
                        item.title?: "",
                        item.body?: "",
                        User(
                            item.user?.avatarUrl?: "",
                            item.user?.login?: ""
                        ),
                        item.htmlUrl?: ""
                    )
                )
            }
        }
        return pullrequests
    }
}
